bdk_rpc_example/src/main.rs

```diff
@@ -0,0 +1,182 @@
+                    RpcData::Blocks(blocks) => {
+                        for (&height, block) in blocks.iter() {
+                            update.insert_checkpoint(BlockId {
+                                height: height as _,
+                                hash: block.block_hash(),
+                            })?;
+                        }
+                        blocks
+                            .into_iter()
+                            .flat_map(|(height, block)| {
+                                block
+                                    .txdata
+                                    .into_iter()
+                                    .map(move |tx| (TxHeight::Confirmed(height as _), tx))
+                            })
+                            .collect::<Vec<_>>()
+                    }
+                    RpcData::Mempool(txs) => txs
+                        .into_iter()
+                        .map(|tx| (TxHeight::Unconfirmed, tx))
+                        .collect::<Vec<_>>(),
+                    RpcData::Stop(finished) => {
+                        log::info!("terminating... sync_finished={}", finished);
+                        drop(recv);
+                        break;
+                    }
+                };
+
+                for (height, tx) in txs {
+                    ensure_spk_cached(&mut keychain_tracker.txout_index, stop_gap);
+
+                    if keychain_tracker.txout_index.is_relevant(&tx) {
+                        update.insert_tx(tx.clone(), Some(height))?;
+                    }
+
+                    for (keychain, this_ind) in keychain_tracker.txout_index.scan(&tx) {
+                        last_active_indexes
+                            .entry(keychain)
+                            .and_modify(|ind| {
+                                if this_ind > *ind {
+                                    *ind = this_ind;
+                                }
+                            })
+                            .or_insert(this_ind);
+                    }
+                }
+
+                let wallet_scan = KeychainScan {
+                    update,
+                    last_active_indexes,
+                };
+
+                let changeset = keychain_tracker.determine_changeset(&wallet_scan)?;
+                db.append_changeset(&changeset)?;
+                keychain_tracker.apply_changeset(changeset);
+            }
+
+            join_handle
+                .join()
+                .expect("failed to join emit_blocks thread")?;
+        }
+    }
+
+    Ok(())
+}
+
+pub fn ensure_spk_cached<K>(txout_index: &mut KeychainTxOutIndex<K>, stop_gap: usize)
+where
+    K: Ord + Clone + core::fmt::Debug,
+{
+    txout_index.trailing_unused_counts().into_iter().for_each(
+        |(keychain, trailing_unused_count)| {
+            let missing_count = stop_gap.saturating_sub(trailing_unused_count);
+            txout_index.derive_new_count(&keychain, missing_count);
+        },
+    );
```

Comment:
I think this should be simpler:

```rust
// for each keychain
let store_up_to = txout_index.last_active_index(&keychain).map(|lai| lai + stop_gap).unwrap_or(stop_gap);
txout_index.store_up_to(derive_up_to);
```

It would be fine to add a helper method to `KeychainTxOutIndex` to make sure all it's keychains have stored up to `last_active_index + stop_gap`.